lib/ui/painting.dart

```diff
@@ -4935,26 +4591,1632 @@ class PictureRecorder extends NativeFieldWrapperClass2 {
+    _drawSkiaPicture(recorder.endRecording() as _SkiaPicture);
+  }
+
+  @override
+  void drawPicture(Picture picture) {
+    if (picture is _SkiaPicture) {
+      _drawSkiaPicture(picture);
+    } else {
+      _drawDisplayListPicture(picture as _DisplayListPicture);
+    }
+  }
+
+  // Constants from physical_shape_layer.cc
+  static const double _kLightHeight = 600;
+  static const double _kLightRadius = 800;
+  Rect _shadowBounds(Rect pathBounds, double elevation) {
+    final double lightRelativeHeight = _kLightHeight - elevation;
+    final double tx = elevation * (_kLightRadius + pathBounds.width * 0.5) / lightRelativeHeight;
+    final double ty = elevation * (_kLightRadius + 600 + pathBounds.height) / lightRelativeHeight;
+    return Rect.fromLTRB(pathBounds.left  - tx, pathBounds.top,
+                         pathBounds.right + tx, pathBounds.bottom + ty);
+  }
+
+  @override
+  void drawShadow(Path path, Color color, double elevation, bool transparentOccluder) {
+    _addOp(transparentOccluder ? _CanvasOp.drawShadowOccluded : _CanvasOp.drawShadow, 2);
+    _addPathData(path);
+    _addInt(color.value);
+    _addScalar(elevation);
+    _addBounds(_shadowBounds(path.getBounds(), elevation), false);
+  }
+}
+
+/// Local storage version of Picture.
+class _DisplayListPicture extends NativeFieldWrapperClass2 implements Picture {
+  _DisplayListPicture._(Rect cullRect,
+                        Rect drawBounds,
+                        Uint8List ops, int numOps,
+                        ByteData data, int numDataBytes,
+                        List<Object> objects)
+      : _cullRect = cullRect,
+        _drawBounds = drawBounds,
+        _ops = ops,
+        _data = data,
+        _objData = objects {
+    _constructor(ops, numOps, data, numDataBytes, objects);
+  }
+  void _constructor(Uint8List ops, int numOps, ByteData data, int dataBytes, List<Object> objects) native 'DisplayList_constructor';
+
+  @override
+  Future<Image> toImage(int width, int height) {
+    if (width <= 0 || height <= 0)
+      throw Exception('Invalid image dimensions.');
+    final Uint8List? ops = _ops;
+    final ByteData? data = _data;
+    final List<Object>? objects = _objData;
+    if (ops == null || data == null || objects == null)
+      throw UnimplementedError('toImage called on disposed Picture');
+    return _futurize(
+          (_Callback<Image?> callback) => _toImage(width, height, (_Image? image) {
+        if (image == null) {
+          callback(null);
+        } else {
+          callback(Image._(image));
+        }
+      }),
+    );
+  }
+
+
+  String? _toImage(int width, int height, _Callback<_Image?> callback) native 'DisplayList_toImage';
+
+  @override
+  void dispose() {
+    _cullRect = null;
+    _drawBounds = null;
+    _ops = null;
+    _data = null;
+    _objData = null;
+  }
+
+  Rect? _cullRect;
+  Rect? _drawBounds;
+  Uint8List? _ops;
+  ByteData? _data;
+  List<Object>? _objData;
```

Comment:
Same comment as above - if we can make this `List<NativeFieldWrapperClass2>` it seems better. Or, perhaps some other private type to `dart:ui` that we can decorate relevant classes with.